### PR TITLE
Do some design

### PR DESCRIPTION
**Pull request by tmds**
This describes down how data will flow and SSH channels implementation hooks into the `SshConnection`

### CODE REVIEW

**Comment by tmds** [comment]
cc @aslicerh @omajid

File: DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+for buffering packets, we can make use of the [Channel](https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channel?view=dotnet-plat-ext-2.1) class.
+
+As a type to represent packets, we want something that works wel together with the `Pipe` related types, like [Memory](https://docs.microsoft.com/en-us/dotnet/api/system.memory-1?view=netstandard-2.1), [ReadOnlySequence<T>](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequence-1?view=netstandard-2.1), and [SequenceReader<T>](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.sequencereader-1?view=netcore-3.0).
+
+We want to pool the packet type so it can be re-used by the `SshConnection`.
+Also, if an instance was parsed, we may want to trim it and allow passing it to the user.
+
+So our packet should be some type of _mutable_ counterpart of `ReadOnlySequence`.
+```cs
+class Sequence
```
Comment (davidfowl) [inline DESIGN.md]: Should this implement `IBufferWriter<byte>`?

File: DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    // Writer API
+    Memory<byte> AllocMemory(int sizeHint); // PipeWriter.GetMemory
+    Span<byte> AllocSpan(int sizeHint);     // PipeWriter.GetSpan
+    void Append(int);                       // PipeWriter.Advance (Adds data put into the Allocated Span/Memory)
+    void Append(Sequence);                  // Extracts data from other sequence and adds it to this sequence.
+    // Reader API
+    void Remove(int);                       // PipeReader.AdvanceTo (Removes data from the front of the Sequence)
+    ReadOnlySequence<byte> GetReadOnlySequence();     // Allows the Sequence to be read, e.g. by SequenceReader
+    ReadOnlySequence<byte> ExtractReadOnlySequence(); // Extracts data from the Sequence
+    // Pooling
```
Comment (davidfowl) [inline DESIGN.md]: These 2 APIs look strange. What's the difference?

File: DESIGN.md (hunk 1 above)
Comment (davidfowl) [reply DESIGN.md]: Also see https://github.com/AArnott/Nerdbank.Streams/blob/master/src/Nerdbank.Streams/Sequence%601.cs

Comment (tmds) [reply DESIGN.md]: > Should this implement IBufferWriter<byte>?

It meets the contract.

> Also see https://github.com/AArnott/Nerdbank.Streams/blob/master/src/Nerdbank.Streams/Sequence%601.cs

This is similar to that type, but tied to `SequencePool` to provide pooling of `Memory`, `Segments`, and the `Sequences` themselves.

File: DESIGN.md (hunk 2 above)
Comment (tmds) [reply DESIGN.md]: If you call the `ExtractReadOnlySequence`, the `Sequence` no longer owns the `Segments` and `Memory`. Next calls to `GetReadOnlySequence` will return an `Empty` `ReadOnlySequence<byte>`. The caller owns the `ReadOnlySequence<byte>` that was handed to him.

**Comment by tmds** [comment]
This is meant to set some direction, merging as-is.

File: DESIGN.md (hunk 2 above)
Comment (davidfowl) [reply DESIGN.md]: I’m not a fan of that subtlety

Comment (tmds) [reply DESIGN.md]: The alternative is to copy data which doesn't need to be copied.

**Pull request merged**